File: app/src/main/java/com/metahumanz/pacilread/reader/modern/content/ReaderPageCache.java
package com.metahumanz.pacilread.reader.modern.content;

import com.metahumanz.pacilread.reader.PageSlice;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

final class ReaderPageCache {
    private final int maxLayoutSignatures;
    private final Map<Integer, List<PageSlice>> activePagesByChapter = new HashMap<>();
    private final Map<ReaderLayoutSignature, Map<Integer, List<PageSlice>>> pagesByLayout;
    private ReaderLayoutSignature activeLayoutSignature;
    private long activeBookId = -1L;

    ReaderPageCache(int maxLayoutSignatures) {
        this.maxLayoutSignatures = Math.max(maxLayoutSignatures, 1);
        this.pagesByLayout = new LinkedHashMap<ReaderLayoutSignature, Map<Integer, List<PageSlice>>>(
                this.maxLayoutSignatures,
                0.75f,
                true
        ) {
            @Override
            protected boolean removeEldestEntry(Map.Entry<ReaderLayoutSignature, Map<Integer, List<PageSlice>>> eldest) {
                return size() > ReaderPageCache.this.maxLayoutSignatures;
            }
        };
    }

    synchronized void setBookId(long bookId) {
        if (activeBookId == bookId) {
            return;
        }
        activeBookId = bookId;
        clear();
    }

    synchronized List<PageSlice> get(int chapterIndex) {
        return activePagesByChapter.get(chapterIndex);
    }

    synchronized boolean contains(int chapterIndex) {
        return activePagesByChapter.containsKey(chapterIndex);
    }

    synchronized void put(int chapterIndex, List<PageSlice> pages) {
        activePagesByChapter.put(chapterIndex, pages);
        rememberActiveLayoutCacheLocked();
    }

    synchronized void clear() {
        activePagesByChapter.clear();
        pagesByLayout.clear();
        activeLayoutSignature = null;
    }

    synchronized boolean hasAny(int chapterIndex) {
        if (activePagesByChapter.containsKey(chapterIndex)) {
            return true;
        }
        for (Map<Integer, List<PageSlice>> pagesByChapter : pagesByLayout.values()) {
            if (pagesByChapter != null && pagesByChapter.containsKey(chapterIndex)) {
                return true;
            }
        }
        return false;
    }

    synchronized boolean activate(ReaderLayoutSignature signature, Runnable beforeIncompatibleSwitch) {
        if (signature == null) {
            return false;
        }
        if (signature.equals(activeLayoutSignature)) {
            return false;
        }
        if (activeLayoutSignature != null && activeLayoutSignature.isPaginationCompatibleWith(signature)) {
            Map<Integer, List<PageSlice>> restored = pagesByLayout.get(signature);
            if (activePagesByChapter.isEmpty() && restored != null) {
                activePagesByChapter.putAll(restored);
            }
            pagesByLayout.remove(activeLayoutSignature);
            activeLayoutSignature = signature;
            rememberActiveLayoutCacheLocked();
            return false;
        }
        if (beforeIncompatibleSwitch != null) {
            beforeIncompatibleSwitch.run();
        }
        rememberActiveLayoutCacheLocked();
        activePagesByChapter.clear();
        Map<Integer, List<PageSlice>> restored = pagesByLayout.get(signature);
        if (restored != null) {
            activePagesByChapter.putAll(restored);
        }
        activeLayoutSignature = signature;
        return true;
    }

    private void rememberActiveLayoutCacheLocked() {
        if (activeLayoutSignature == null) {
            return;
        }
        if (activePagesByChapter.isEmpty()) {
            pagesByLayout.remove(activeLayoutSignature);
        } else {
            pagesByLayout.put(activeLayoutSignature, new HashMap<>(activePagesByChapter));
        }
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/reader/modern/content/ReaderPageCache.kt
package com.metahumanz.pacilread.reader.modern.content

import com.metahumanz.pacilread.reader.PageSlice
import java.util.LinkedHashMap

class ReaderPageCache(maxLayoutSignatures: Int) {
    private val maxLayoutSignatures = maxLayoutSignatures.coerceAtLeast(1)
    private val activePagesByChapter: MutableMap<Int, List<PageSlice>> = HashMap()
    private val pagesByLayout: MutableMap<ReaderLayoutSignature, MutableMap<Int, List<PageSlice>>> =
        object : LinkedHashMap<ReaderLayoutSignature, MutableMap<Int, List<PageSlice>>>(this.maxLayoutSignatures, 0.75f, true) {
            override fun removeEldestEntry(eldest: MutableMap.MutableEntry<ReaderLayoutSignature, MutableMap<Int, List<PageSlice>>>?): Boolean =
                size > this@ReaderPageCache.maxLayoutSignatures
        }
    private var activeLayoutSignature: ReaderLayoutSignature? = null
    private var activeBookId = -1L

    @Synchronized
    fun setBookId(bookId: Long) {
        if (activeBookId == bookId) return
        activeBookId = bookId
        clear()
    }

    @Synchronized
    fun get(chapterIndex: Int): List<PageSlice>? = activePagesByChapter[chapterIndex]

    @Synchronized
    fun contains(chapterIndex: Int): Boolean = activePagesByChapter.containsKey(chapterIndex)

    @Synchronized
    fun put(chapterIndex: Int, pages: List<PageSlice>) {
        activePagesByChapter[chapterIndex] = pages
        rememberActiveLayoutCacheLocked()
    }

    @Synchronized
    fun clear() {
        activePagesByChapter.clear()
        pagesByLayout.clear()
        activeLayoutSignature = null
    }

    @Synchronized
    fun hasAny(chapterIndex: Int): Boolean {
        if (activePagesByChapter.containsKey(chapterIndex)) return true
        for (pagesByChapter in pagesByLayout.values) {
            if (pagesByChapter.containsKey(chapterIndex)) return true
        }
        return false
    }

    @Synchronized
    fun activate(signature: ReaderLayoutSignature?, beforeIncompatibleSwitch: Runnable?): Boolean {
        if (signature == null) return false
        if (signature == activeLayoutSignature) return false
        val activeSignature = activeLayoutSignature
        if (activeSignature != null && activeSignature.isPaginationCompatibleWith(signature)) {
            val restored = pagesByLayout[signature]
            if (activePagesByChapter.isEmpty() && restored != null) activePagesByChapter.putAll(restored)
            pagesByLayout.remove(activeSignature)
            activeLayoutSignature = signature
            rememberActiveLayoutCacheLocked()
            return false
        }
        beforeIncompatibleSwitch?.run()
        rememberActiveLayoutCacheLocked()
        activePagesByChapter.clear()
        pagesByLayout[signature]?.let { activePagesByChapter.putAll(it) }
        activeLayoutSignature = signature
        return true
    }

    private fun rememberActiveLayoutCacheLocked() {
        val signature = activeLayoutSignature ?: return
        if (activePagesByChapter.isEmpty()) {
            pagesByLayout.remove(signature)
        } else {
            pagesByLayout[signature] = HashMap(activePagesByChapter)
        }
    }
}
